### ISSUE TITLE
Issue importing osgeo under Miniconda Python 3.10

### ISSUE THREAD

**Issue by milos-u** [OSGeo/gdal#5623]
Trying to import osgeo fails:

```(p310_x32) python
Python 3.10.4 | packaged by conda-forge | (main, Mar 30 2022, 13:41:19) [MSC v.1
916 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import osgeo
Traceback (most recent call last):
  File "C:\ProgramData\Miniconda3\envs\p310_x32\lib\site-packages
\osgeo\__init__.py", line 29, in swig_import_helper
    return importlib.import_module(mname)
  File "C:\ProgramData\Miniconda3\envs\p310_x32\lib\importlib\__i
nit__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 674, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 571, in module_from_spec
  File "<frozen importlib._bootstrap_external>", line 1176, in create_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
ImportError: DLL load failed while importing _gdal: The specified procedure could not be found.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\ProgramData\Miniconda3\envs\p310_x32\lib\site-packages
\osgeo\__init__.py", line 45, in <module>
    _gdal = swig_import_helper()
  File "C:\ProgramData\Miniconda3\envs\p310_x32\lib\site-packages
\osgeo\__init__.py", line 41, in swig_import_helper
    raise ImportError(traceback_string + '\n' + msg)
ImportError: Traceback (most recent call last):
  File "C:\ProgramData\Miniconda3\envs\p310_x32\lib\site-packages
\osgeo\__init__.py", line 29, in swig_import_helper
    return importlib.import_module(mname)
  File "C:\ProgramData\Miniconda3\envs\p310_x32\lib\importlib\__i
nit__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 674, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 571, in module_from_spec
  File "<frozen importlib._bootstrap_external>", line 1176, in create_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
ImportError: DLL load failed while importing _gdal: The specified procedure could not be found.

On Windows, with Python >= 3.8, DLLs are no longer imported from the PATH.
If gdalXXX.dll is in the PATH, then set the USE_PATH_FOR_GDAL_PYTHON=YES environ
ment variable
to feed the PATH into os.add_dll_directory().
>>>
```

I have some directories in PATH that contain different versions of GDAL (I know removing them from PATH would fix this issue).

Looking into osgeo/__init__.py I can see that these directories are added into Python DLL search path.

One would expect that everything would be working OK, because GDAL Library/bin/ directory is added by conda "activate" as one of the first items to the PATH.

The problem is, when looking into Sysinternals ProcMon, that the DLL directories are searched in reverse order, i.e. the DLL directory from the last call to os.add_dll_directory() is searched first.

Miniconda will add GDAL Library directory at the begining of the search PATH and the result is that this directory is searched as the last one and thus Windows will load DLLs from another directory in PATH, containing different version of GDAL.

Setting USE_PATH_FOR_GDAL_PYTHON=1 does not help either because the DLL search path is still in the reverse order.

The following patch would fix the issue in my case (with this patch in place there is also no need to set USE_PATH_FOR_GDAL_PYTHON).

```(p310_x32) C:\ProgramData\Miniconda3\envs\p310_x32\Lib\site-packages\osgeo>diff -u __init__.py  __init_.py.patched
--- __init__.py 2022-04-17 12:01:33.083212000 +0200
+++ __init_.py.patched  2022-04-17 12:35:28.155749800 +0200
@@ -5,7 +5,7 @@
 if version_info >= (3, 8, 0) and platform == 'win32':
     import os
     if 'USE_PATH_FOR_GDAL_PYTHON' in os.environ and 'PATH' in os.environ:
-        for p in os.environ['PATH'].split(';'):
+        for p in os.environ['PATH'].split(';')[::-1]:
             if p:
                 try:
                     os.add_dll_directory(p)
@@ -13,7 +13,7 @@
                     continue
     elif 'PATH' in os.environ:
         import glob
-        for p in os.environ['PATH'].split(';'):
+        for p in os.environ['PATH'].split(';')[::-1]:
             if glob.glob(os.path.join(p, 'gdal*.dll')) or glob.glob(os.path.joi
n(p, 'libgdal*.dll')):
                 try:
                     os.add_dll_directory(p)
```

Environment:
```
gdalinfo --version
GDAL 3.4.1, released 2021/12/27

python -V
Python 3.10.4

conda -V
conda 4.11.0

OS: Windows 2012 R2 Foundation
```

**Comment by rouault** [comment]
> The problem is, when looking into Sysinternals ProcMon, that the DLL directories are searched in reverse order, i.e. the DLL directory from the last call to os.add_dll_directory() is searched first.

Unfortunately https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-adddlldirectory mentions "If AddDllDirectory is used to add more than one directory to the process DLL search path, the order in which those directories are searched is unspecified."
Perhaps we should just stop at the first directory in the PATH that matches "glob.glob(os.path.join(p, 'gdal*.dll')) or glob.glob(os.path.join(p, 'libgdal*.dll'))"

**Comment by rouault** [comment]
CC @sgillies as I see that https://github.com/rasterio/rasterio/blob/master/rasterio/_loading.py behaves the same as GDAL's Python bindings loader

**Comment by milos-u** [comment]
> > The problem is, when looking into Sysinternals ProcMon, that the DLL directories are searched in reverse order, i.e. the DLL directory from the last call to os.add_dll_directory() is searched first.
> 
> Unfortunately https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-adddlldirectory mentions "If AddDllDirectory is used to add more than one directory to the process DLL search path, the order in which those directories are searched is unspecified." Perhaps we should just stop at the first directory in the PATH that matches "glob.glob(os.path.join(p, 'gdal*.dll')) or glob.glob(os.path.join(p, 'libgdal*.dll'))"

You're correct, the following patch works in my case as well:

```
--- __init__.py 2022-04-17 13:44:41.599116800 +0200
+++ __init__.py.patched 2022-04-17 13:45:12.693729600 +0200
@@ -17,6 +17,7 @@
             if glob.glob(os.path.join(p, 'gdal*.dll')) or glob.glob(os.path.joi
n(p, 'libgdal*.dll')):
                 try:
                     os.add_dll_directory(p)
+                    break
                 except (FileNotFoundError, OSError):
                     continue

```

**Comment by zuhlmann** [comment]
Hi.  Hopefully this is the correct venue to post this question.  I am seeing the same behavior.

I have the same specs: 
GDAL 3.4.1 
Python 3.10.4 
Conda 4.11.0 
Windows 11, along with a Conda Environment with only the **qgis** packages installed from conda-forge.

When I try anything relying on the GDAL.dll, in this case, 
`>>>python  %CONDA_PREFIX%\Scripts\gdal_calc.py...`
I get the same traceback as the original post

`...
On Windows, with Python >= 3.8, DLLs are no longer imported from the PATH.
If gdalXXX.dll is in the PATH, then set the USE_PATH_FOR_GDAL_PYTHON=YES environ
ment variable
to feed the PATH into os.add_dll_directory().`

Additionally, directly running a gdal command such as `gdalinfo` in the command yields this:
![Screenshot 2022-05-13 101830](https://user-images.githubusercontent.com/48263393/168327643-722c8654-246e-4317-befd-e6cc539e3966.png)
